### DISCUSSION TITLE
zls-only Type Hint

### DISCUSSION THREAD

**Discussion by iacore** [zigtools/zls#1506] [Ideas]
I have this idea that will make zls provide auto-completion for `anytype`. 

```zig
pub fn foo(writer: anytype) void {
    // @zls-type writer: std.io.Writer(*anyopaque, error{}, undefined)
    // this will force zls to think `writer` is that type
}
```

**Comment by techatrix** [comment]
ZLS can resolve the types of an anytype parameter by collecting all callers of that function which has been implemented in #1067.

Here is a code example that shows this feature in action:
```zig
const std = @import("std");

fn callerWriter() !void {
    const stdout = std.io.getStdOut().writer();
    calleeWriter(stdout);
}

fn calleeWriter(writer: anytype) void {
    writer. // ask for completions here
}

fn callerArray() !void {
    const stdout = std.ArrayListUnmanaged(u8){};
    calleeArray(stdout);
}

fn calleeArray(array: anytype) void {
    array. // ask for completions here
}
```

As far as I know, this feature has an issue with how it collects callers across different files which is why this may not always appear to work.

**Reply by iacore** [reply]
What else have I missed out on?

**Comment by highjeans** [comment]
I have a use case for this (I know it is years later). I am using microzig to do some embedded development, and if when I use the compile-time pin configuration it allows me to do, then running `pin_config.apply()` makes zls return that its type is `unknown`, preventing any completions on it. With type hints, I can get completions easily rather than me having to look at the source to figure out its type and what functions it supports. I might try and implement this myself but currently I am stuck on zig 0.15.2 until microzig supports zig 0.16.x.
